### PR TITLE
Adding SOC Occupation mapping to Stat Var Dcid Generator

### PR DESCRIPTION
**Pull request by Abilityguy**
SOC codes to occupation mapping added to dcid generator.
Added tests for the same.
`test_naics_name_generation` has been defined twice in `statvar_dcid_generator_test.py`(line 390 and 260). Removed one definition.

### CODE REVIEW

File: util/statvar_dcid_generator.py [hunk 1]
```diff
@@ -342,6 +355,32 @@ def _process_naics_range(range_str: str) -> str:
...
+    if soc_val:
+        processed_str = 'SOC'
+
+        # Remove namespaces
+        soc_val = soc_val[soc_val.find(':') + 1:]
+
+        # Strip SOCv2018/
+        soc_val = soc_val.replace('SOCv2018/', '')
+
+        processed_str = processed_str + SOC_MAP[soc_val]
```
Comment (ajaits) [inline util/statvar_dcid_generator.py:379]: What if soc_val is not present in SOC_MAP?

Comment (Abilityguy) [reply util/statvar_dcid_generator.py]: Done. Have pushed the fix.

File: util/statvar_dcid_generator_test.py [hunk 2]
```diff
@@ -257,56 +257,6 @@ def test_measurement_denominator(self):
                          'AsAFractionOf_Count_Person_25To64Years')
         self.assertEqual(dcid, expected_dcid)
 
-    def test_naics_name_generation(self):
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/statvar_dcid_generator_test.py:260]: I looked at the docs and couldn't find any such option.

File: util/statvar_dcid_generator.py [hunk 3]
```diff
@@ -342,6 +355,32 @@ def _process_naics_range(range_str: str) -> str:
...
+        is empty.
+    """
+    if soc_val:
+        processed_str = 'SOC'
+
+        # Remove namespaces
+        soc_val = soc_val[soc_val.find(':') + 1:]
+
+        # Strip SOCv2018/
+        soc_val = soc_val.replace('SOCv2018/', '')
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/statvar_dcid_generator.py]: Done. Prefix check from below comment solves this.

File: util/statvar_dcid_generator.py [hunk 4]
```diff
@@ -404,6 +443,8 @@ def _process_constraint_property(prop: str, value: str) -> str:
     """
     if prop == 'naics':
         name = _naics_code_to_name(value)
+    elif prop == 'occupation':
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/statvar_dcid_generator.py]: Done. Checking for prefix 'SOCv2018' instead.

File: util/soc_codes_names.py [hunk 5]
```diff
@@ -0,0 +1,2975 @@
...
+    soc_map = dict()
+    soc_mcf = requests.get(
+        'https://raw.githubusercontent.com/datacommonsorg/schema/main/core/soc.mcf'
+    ).text
+    soc_mcf = soc_mcf.split('\n')
+
+    for line in soc_mcf:
+        # Extracting the SOC code from the dcid
+        # The dcid is of the form 'dcid:SOCv2018/<code>'
+        if line.startswith('Node:'):
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py:36]: Done.

File: util/soc_codes_names.py [hunk 6]
```diff
@@ -0,0 +1,2975 @@
...
+            code = line.split('SOCv2018/')[1]
+
+        # Extracting occupation from name
+        if line.startswith('name'):
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py]: Done. Have added a check.

File: util/soc_codes_names.py [hunk 7]
```diff
@@ -0,0 +1,2975 @@
...
+            occupation = line.split(':')[1].strip()
+            occupation = soc_name_clean_regex.sub('', occupation)
+            soc_map[code] = occupation  # Assign occupation to code from dcid
+
+    return soc_map
+
+
+SOC_MAP = {
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py:55]: Done.

File: util/soc_codes_names.py [hunk 8]
```diff
@@ -0,0 +1,2975 @@
...
+    '47-2230':
+        'SolarPhotovoltaicInstallers',
+    '47-2231':
+        'SolarPhotovoltaicInstallers',
+    '47-3000':
+        'HelpersConstructionTrades',
+    '47-3010':
+        'HelpersConstructionTrades',
+    '47-3011':
+        'Helpers--BrickmasonsBlockmasonsStonemasonsTileMarbleSetters',
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py]: Done.

File: util/soc_codes_names.py [hunk 9]
```diff
@@ -0,0 +1,2975 @@
...
+        'Service',
+    'highLevelAggregation-3':
+        'SalesOffice',
+    'highLevelAggregation-4':
+        'NaturalResourcesConstructionMaintenance',
+    'highLevelAggregation-5':
+        'ProductionTransportationMaterialMoving',
+    'highLevelAggregation-6':
+        'MilitarySpecific'
+}
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py:2982]: Done. Added a test that checks for extra characters.
A manual check was made for the duplicate values. Keys with duplicate values share a common prefix.

File: util/soc_codes_names.py [hunk 10]
```diff
@@ -0,0 +1,2975 @@
...
+        # Extracting the SOC code from the dcid
+        # The dcid is of the form 'dcid:SOCv2018/<code>'
+        if line.startswith('Node:'):
+            code = line.split('SOCv2018/')[1]
+
+        # Extracting occupation from name
+        if line.startswith('name'):
+            occupation = line.split(':')[1].strip()
+            occupation = soc_name_clean_regex.sub('', occupation)
+            soc_map[code] = occupation  # Assign occupation to code from dcid
```
[the remark this answers is not in the document]
Comment (Abilityguy) [reply util/soc_codes_names.py:50]: Done.

File: util/statvar_dcid_generator.py [hunk 11]
```diff
@@ -204,6 +217,12 @@
     },
     'placeOfBirth': {
         'prepend': 'PlaceOfBirth'
+    },
```
Comment (Abilityguy) [inline util/statvar_dcid_generator.py:220]: Added new entries to PREPEND_APPEND_REPLACE Map to aid current imports.

File: util/soc_codes_names.py [hunk 12]
```diff
@@ -0,0 +1,2985 @@
...
+    soc_mcf = requests.get(
+        'https://raw.githubusercontent.com/datacommonsorg/schema/main/core/soc.mcf'
+    ).text
+    soc_mcf = soc_mcf.split('\n')
+
+    for line in soc_mcf:
+        # Extracting the SOC code from the dcid
+        # The dcid is of the form 'dcid:SOCv2018/<code>'
+        if line.startswith('Node:'):
+            if 'SOCv2018' in line:
```
Comment (ajaits) [inline util/soc_codes_names.py:37]: if 'SOCv2018/' in line:
  so split()[1] is always present

File: util/statvar_dcid_generator.py [hunk 13]
```diff
@@ -404,6 +460,8 @@ def _process_constraint_property(prop: str, value: str) -> str:
     """
     if prop == 'naics':
         name = _naics_code_to_name(value)
+    elif 'SOCv2018' in value:
```
Comment (ajaits) [inline util/statvar_dcid_generator.py:463]: 'SOCv2018/' in value:

File: util/soc_codes_names.py (hunk 12 above)
Comment (Abilityguy) [reply util/soc_codes_names.py]: Done.

File: util/statvar_dcid_generator.py (hunk 13 above)
Comment (Abilityguy) [reply util/statvar_dcid_generator.py]: Done.

**Review by ajaits — APPROVED** [review]

**Review by ajaits — APPROVED** [review]

**Pull request merged**